source/client-side-encryption/client-side-encryption.rst

```diff
@@ -1332,6 +1367,67 @@ on sockets. See the SDAM spec description of `cooldownMS <../source/server-disco
 Because single threaded drivers may exceed ``serverSelectionTimeoutMS`` by the
 duration of the topology scan, ``connectTimeoutMS`` is also reduced.
 
+What's the deal with metadataClient, keyVaultClient, and the internal client?
+-----------------------------------------------------------------------------
+
+When automatically encrypting a command, the driver runs:
+- a ``listCollections`` command to determine if the target collection
+has a remote schema. This uses the ``metadataClient``.
+- a ``find`` against the key vault collection to fetch keys. This uses the
+``keyVaultClient``.
+
+**Why not reuse the parent MongoClient?**
+
+These operations MUST NOT reuse the same connection pool as the parent
+``MongoClient`` configured with automatic encryption to avoid possible deadlock
+situations.
+
+Drivers supporting a connection pool (see `CMAP specification
+</source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst>`_)
+support an option for limiting the connection pool size: ``maxPoolSize``.
+
+Drivers need to check out a connection before serializing the command. If the
+``listCollections`` or ``find`` command during automatic encryption uses the same
+connection pool as the parent MongoClient, the application is susceptible to
+deadlocks.
```

Comment:
It it, technically, a deadlock given that the checkout will eventually time out?

Reply:
If there is a `waitQueueTimeoutMS` set, then it would be a timeout instead of a deadlock. The default behavior is a deadlock though, since `waitQueueTimeoutMS` defaults to "no limit": https://github.com/mongodb/specifications/blob/master/source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst#connection-pool-options-1